File: tfcc_code_generator/frontend/tensorflow/converter/nn/conv2d.py
# Copyright 2021 Wechat Group, Tencent
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import numpy as np
from tensorflow.python.framework.ops import Operation
import ir.framework
import ir.node
from ir.common import create_constant_symbol
from ..converter import Converter


# implement for https://tensorflow.google.cn/api_docs/python/tf/raw_ops/Conv2D?hl=en
class Conv2D(Converter):
    def get_padding_size(self, input_size, stride_size, kernel_size):
        if stride_size == 1:
            return kernel_size - 1
        else:
            if not isinstance(input_size, int):
                return -1
            output_size = (input_size + stride_size - 1) // 2
            padding = (output_size - 1) * stride_size + kernel_size - input_size
            padding = max(0, padding)
            return padding

    def __call__(self, op: Operation, graph: ir.framework.Graph):
        if not self.accept(op):
            return False

        inp_strs = [inp.name for inp in op.inputs]
        oup_strs = [oup.name for oup in op.outputs]

        assert len(inp_strs) == 2
        assert len(oup_strs) == 1

        input_symbol = graph.get_symbol(inp_strs[0])
        kernel_symbol = graph.get_symbol(inp_strs[1])

        data_format = op.get_attr("data_format")
        strides = op.get_attr("strides")
        padding = op.get_attr("padding")
        dilations = op.get_attr("dilations")

        if data_format == b"NHWC":
            n, h, w, c = input_symbol.shape
            nchw = False
        elif data_format == b"NCHW":
            n, c, h, w = input_symbol.shape
            nchw = True
        else:
            return False

        kernel_height = kernel_symbol.shape[0]
        kernel_width = kernel_symbol.shape[1]

        if len(strides) != 4:
            return False
        if nchw:
            if strides[0] != 1 or strides[1] != 1:
                return False
            stride_height = strides[2]
            stride_width = strides[3]
        else:
            if strides[0] != 1 or strides[3] != 1:
                return False
            stride_height = strides[1]
            stride_width = strides[2]

        if len(dilations) != 4:
            return False
        if nchw:
            if dilations[0] != 1 or dilations[1] != 1:
                return False
            dilation_height = dilations[2]
            dilation_width = dilations[3]
        else:
            if dilations[0] != 1 or dilations[3] != 1:
                return False
            dilation_height = dilations[1]
            dilation_width = dilations[2]

        real_input_name = inp_strs[0]

        if padding == b"VALID":
            padding_height = 0
            padding_width = 0
        elif padding == b"SAME":
            if dilation_height != 1 or dilation_width != 1:
                return False
            if not isinstance(kernel_height, int) or not isinstance(kernel_width, int):
                return False

            total_padding_height = self.get_padding_size(
                h, stride_height, kernel_height
            )
            total_padding_width = self.get_padding_size(w, stride_width, kernel_width)
            if total_padding_height < 0 or total_padding_width < 0:
                return False
            if total_padding_height % 2 == 0:
                padding_height = total_padding_height // 2
            else:
                padding_output_name = graph.context.create_symbol_name(real_input_name)
                padding_head_name = graph.context.create_symbol_name(
                    "{}_padding_head".format(inp_strs[0])
                )
                create_constant_symbol(
                    graph,
                    padding_head_name,
                    ir.framework.DataType.UINT32,
                    ir.framework.SymbolType.CONSTANT_VALUE,
                    [1],
                    np.asarray([total_padding_height // 2], dtype=np.uint32),
                )
                padding_tail_name = graph.context.create_symbol_name(
                    "{}_padding_tail".format(inp_strs[0])
                )
                create_constant_symbol(
                    graph,
                    padding_tail_name,
                    ir.framework.DataType.UINT32,
                    ir.framework.SymbolType.CONSTANT_VALUE,
                    [1],
                    np.asarray([total_padding_height // 2 + 1], dtype=np.uint32),
                )
                if data_format == b"NHWC":
                    axis = 1
                else:
                    axis = 2
                graph.append_node(
                    ir.node.base.Pad(
                        op.name + ":0",
                        graph,
                        [real_input_name, padding_head_name, padding_tail_name],
                        [padding_output_name],
                        axis=axis,
                    )
                )
                real_input_name = padding_output_name
                padding_height = 0
            if total_padding_width % 2 == 0:
                padding_width = total_padding_width // 2
            else:
                padding_output_name = graph.context.create_symbol_name(real_input_name)
                padding_head_name = graph.context.create_symbol_name(
                    "{}_padding_head".format(inp_strs[0])
                )
                create_constant_symbol(
                    graph,
                    padding_head_name,
                    ir.framework.DataType.UINT32,
                    ir.framework.SymbolType.CONSTANT_VALUE,
                    [1],
                    np.asarray([total_padding_width // 2], dtype=np.uint32),
                )
                padding_tail_name = graph.context.create_symbol_name(
                    "{}_padding_tail".format(inp_strs[0])
                )
                create_constant_symbol(
                    graph,
                    padding_tail_name,
                    ir.framework.DataType.UINT32,
                    ir.framework.SymbolType.CONSTANT_VALUE,
                    [1],
                    np.asarray([total_padding_width // 2 + 1], dtype=np.uint32),
                )
                if data_format == b"NHWC":
                    axis = 2
                else:
                    axis = 3
                graph.append_node(
                    ir.node.base.Pad(
                        op.name + ":1",
                        graph,
                        [real_input_name, padding_head_name, padding_tail_name],
                        [padding_output_name],
                        axis=axis,
                    )
                )
                real_input_name = padding_output_name
                padding_width = 0
        else:
            return False

        kernel_transpose_output_name = graph.context.create_symbol_name(inp_strs[1])
        graph.append_node(
            ir.node.base.Transpose(
                op.name + ":2",
                graph,
                [inp_strs[1]],
                [kernel_transpose_output_name],
                perm=[3, 2, 0, 1],
            )
        )
        attrs = {
            "padding_height": padding_height,
            "padding_width": padding_width,
            "stride_height": stride_height,
            "stride_width": stride_width,
            "dilate_height": dilation_height,
            "dilate_width": dilation_width,
            "group": 1,
            "nchw": nchw,
        }
        graph.append_node(
            ir.node.nn.Conv(
                op.name + ":3",
                graph,
                [real_input_name, kernel_transpose_output_name],
                oup_strs,
                **attrs
            )
        )
        return True

File: tfcc_runtime/operations/nn/conv2d.cpp
// Copyright 2021 Wechat Group, Tencent
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "conv2d.h"

#include "tfcc_runtime/framework/graph.h"
#include "tfcc_runtime/proto/operations/nn.pb.h"
#include "tfcc_runtime/utils/commutils.h"

namespace tfcc {
namespace runtime {
namespace nn {

template <class T>
static const char* _wrapper_conv2d_nchw(
    const tfcc::Tensor<T>* a, const tfcc::Tensor<T>* kernel, unsigned paddingHeight,
    unsigned paddingWidth, unsigned strideHeight, unsigned strideWidth, unsigned dilateHeight,
    unsigned dilateWidth, tfcc::Variable<T>* result) noexcept {
  try {
    *result = tfcc::nn::conv2d(
        *a, false, *kernel, paddingHeight, paddingWidth, strideHeight, strideWidth, dilateHeight,
        dilateWidth);
  } catch (std::exception& e) {
    thread_local std::string reason = e.what();
    return reason.c_str();
  }
  return nullptr;
}

template <class T>
static const char* _wrapper_conv2d_nhwc(
    const tfcc::Tensor<T>* a, const tfcc::Tensor<T>* kernel, unsigned paddingHeight,
    unsigned paddingWidth, unsigned strideHeight, unsigned strideWidth, unsigned dilateHeight,
    unsigned dilateWidth, tfcc::Variable<T>* result) noexcept {
  try {
    *result = tfcc::nn::conv2d(
        *a, true, *kernel, paddingHeight, paddingWidth, strideHeight, strideWidth, dilateHeight,
        dilateWidth);
  } catch (std::exception& e) {
    thread_local std::string reason = e.what();
    return reason.c_str();
  }
  return nullptr;
}

template <class T>
static const char* _wrapper_conv2d_nchw_with_group(
    const tfcc::Tensor<T>* a, const tfcc::Tensor<T>* kernel, unsigned paddingHeight,
    unsigned paddingWidth, unsigned strideHeight, unsigned strideWidth, unsigned dilateHeight,
    unsigned dilateWidth, unsigned group, tfcc::Variable<T>* result) noexcept {
  try {
    *result = tfcc::nn::conv2d(
        *a, false, *kernel, paddingHeight, paddingWidth, strideHeight, strideWidth, dilateHeight,
        dilateWidth, group);
  } catch (std::exception& e) {
    thread_local std::string reason = e.what();
    return reason.c_str();
  }
  return nullptr;
}

template <class T>
static const char* _wrapper_conv2d_nhwc_with_group(
    const tfcc::Tensor<T>* a, const tfcc::Tensor<T>* kernel, unsigned paddingHeight,
    unsigned paddingWidth, unsigned strideHeight, unsigned strideWidth, unsigned dilateHeight,
    unsigned dilateWidth, unsigned group, tfcc::Variable<T>* result) noexcept {
  try {
    *result = tfcc::nn::conv2d(
        *a, true, *kernel, paddingHeight, paddingWidth, strideHeight, strideWidth, dilateHeight,
        dilateWidth, group);
  } catch (std::exception& e) {
    thread_local std::string reason = e.what();
    return reason.c_str();
  }
  return nullptr;
}

template <tfcc::runtime::common::DataType DTYPE>
std::string Conv2D<DTYPE>::getOperationTypeUrl() const {
  tfcc::runtime::operations::nn::Conv2D operation;
  return get_protobuf_type_url(operation);
}

template <tfcc::runtime::common::DataType DTYPE>
std::set<unsigned> Conv2D<DTYPE>::getOperationVersions() const {
  std::set<unsigned> versions = {
      tfcc::runtime::operations::nn::Conv2D::VERSION_1,
      tfcc::runtime::operations::nn::Conv2D::VERSION_2,
      tfcc::runtime::operations::nn::Conv2D::VERSION_3,
      tfcc::runtime::operations::nn::Conv2D::VERSION_4,
  };
  return versions;
}

template <tfcc::runtime::common::DataType DTYPE>
std::unique_ptr<OperationResource> Conv2D<DTYPE>::process(
    const tfcc::runtime::model::Node& node, Graph& graph) const {
  if (node.inputs_size() != 2 || node.outputs_size() != 1) {
    return nullptr;
  }
  if (!node.operation().Is<tfcc::runtime::operations::nn::Conv2D>()) {
    return nullptr;
  }
  auto& symbolManager = getSymbolManager(graph);
  if (symbolManager.getSymbolInfo(node.outputs(0)).dtype != DTYPE) {
    return nullptr;
  }
  if (symbolManager.getSymbolInfo(node.inputs(0)).dtype != DTYPE ||
      symbolManager.getSymbolInfo(node.inputs(1)).dtype != DTYPE) {
    return nullptr;
  }

  std::unique_ptr<OperationResource> resource(new OperationResource());
  tfcc::runtime::operations::nn::Conv2D operation;
  node.operation().UnpackTo(&operation);

  auto& jit = getJIT(graph);

  const tfcc::Tensor<T>* inputSymbol = symbolManager.getTensor(node.inputs(0), T());
  const tfcc::Tensor<T>* kernelSymbol = symbolManager.getTensor(node.inputs(1), T());
  tfcc::Variable<T>* outputSymbol = symbolManager.getVariable(node.outputs(0), T());

  unsigned dilate_height = operation.dilate_height();
  unsigned dilate_width = operation.dilate_width();
  dilate_height = dilate_height == 0 ? 1 : dilate_height;
  dilate_width = dilate_width == 0 ? 1 : dilate_width;
  bool nchw = true;
  if (node.version() != tfcc::runtime::operations::nn::Conv2D::VERSION_1 &&
      node.version() != tfcc::runtime::operations::nn::Conv2D::VERSION_2) {
    nchw = operation.nchw();
  }
  uint32_t group = 1;
  if (operation.group() != 0) {
    group = operation.group();
  }
  if (group == 1) {
    if (nchw) {
      callFunction(
          jit, _wrapper_conv2d_nchw<T>, inputSymbol, kernelSymbol, operation.padding_height(),
          operation.padding_width(), operation.stride_height(), operation.stride_width(),
          dilate_height, dilate_width, outputSymbol);
    } else {
      callFunction(
          jit, _wrapper_conv2d_nhwc<T>, inputSymbol, kernelSymbol, operation.padding_height(),
          operation.padding_width(), operation.stride_height(), operation.stride_width(),
          dilate_height, dilate_width, outputSymbol);
    }
  } else {
    if (nchw) {
      callFunction(
          jit, _wrapper_conv2d_nchw_with_group<T>, inputSymbol, kernelSymbol,
          operation.padding_height(), operation.padding_width(), operation.stride_height(),
          operation.stride_width(), dilate_height, dilate_width, group, outputSymbol);
    } else {
      callFunction(
          jit, _wrapper_conv2d_nhwc_with_group<T>, inputSymbol, kernelSymbol,
          operation.padding_height(), operation.padding_width(), operation.stride_height(),
          operation.stride_width(), dilate_height, dilate_width, group, outputSymbol);
    }
  }
  return resource;
}

std::vector<std::unique_ptr<Operation>> get_conv2d_operations() {
#define DEFINE_FUNC(dtype) operations.emplace_back(std::unique_ptr<Operation>(new Conv2D<dtype>()));
  std::vector<std::unique_ptr<Operation>> operations;
  TFCC_RUNTIME_FOR_ALL_DATA_TYPE(DEFINE_FUNC);
  return operations;
}

}  // namespace nn
}  // namespace runtime
}  // namespace tfcc
